src/pybind/ctc/ctc_pybind.cc

```diff
@@ -0,0 +1,215 @@
+// pybind/ctc/ctc_pybind.cc
+
+// Copyright 2020   Mobvoi AI Lab, Beijing, China
+//                  (author: Fangjun Kuang, Yaguang Hu, Jian Wang)
+
+// See ../../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ctc/ctc_pybind.h"
+
+#include "ctc/warp-ctc/include/ctc.h"
+#include "dlpack/dlpack_subvector.h"
+
+using namespace kaldi;
+
+namespace {
+
+// note that the argument order does not follow the google style
+// which puts the output argument last;
+ctcStatus_t compute_ctc_loss_cpu(const DLPackSubVector<float>& activations,
+                                 DLPackSubVector<float>* gradients,
+                                 const DLPackSubVector<int>& flat_labels,
+                                 const DLPackSubVector<int>& label_lengths,
+                                 const DLPackSubVector<int>& input_lengths,
+                                 int alphabet_size, int minibatch,
+                                 DLPackSubVector<float>* costs,
+                                 DLPackSubVector<float>* workspace,
+                                 ctcOptions options) {
+  ctcStatus_t status = compute_ctc_loss(
+      activations.Data(), gradients->Data(), flat_labels.Data(),
+      label_lengths.Data(), input_lengths.Data(), alphabet_size, minibatch,
+      costs->Data(), workspace->Data(), options);
+  return status;
+}
+
+ctcStatus_t compute_ctc_loss_gpu(const DLPackCuSubVector<float>& activations,
+                                 DLPackCuSubVector<float>* gradients,
+                                 const DLPackSubVector<int>& flat_labels,
+                                 const DLPackSubVector<int>& label_lengths,
+                                 const DLPackSubVector<int>& input_lengths,
+                                 int alphabet_size, int minibatch,
+                                 DLPackSubVector<float>* costs,
+                                 DLPackCuSubVector<float>* workspace,
+                                 ctcOptions options) {
+  options.stream = nullptr;  // use the default stream
+  ctcStatus_t status = compute_ctc_loss(
+      activations.Data(), gradients->Data(), flat_labels.Data(),
+      label_lengths.Data(), input_lengths.Data(), alphabet_size, minibatch,
+      costs->Data(), workspace->Data(), options);
+  return status;
+
+  // we use the default CUDA stream by setting it to nullptr, i.e., 0
```

Comment:
This comment is too far from the code it refers to.  Also, how would this interact with multi-GPU training?  Isn't that a common scenario?  And right now, how would our Kaldi LF-MMI code deal with that?

Reply:
This function is invoked inside
https://github.com/kaldi-asr/kaldi/pull/3834/commits/5ae7abc1907cc69014674c345cbb8d817bafe209#diff-a3c9284f584c43207ad2eb032edbfee6R88

so we can always use the stream corresponding to the current GPU device selected in PyTorch.

I think it can be scaled without any modifications to mutiple GPUs.

Reply:
I thought users will normally not want to see the implementation details so
I put the comment at the end.

Should I move it to the beginning ?

Reply:
fixed. Moved to the beginning.